core/src/main/java/de/danoeh/antennapod/core/storage/DBWriter.java

```diff
@@ -136,6 +130,21 @@ private DBWriter() {
         });
     }
 
+    /**
+     * Deletes a downloaded FeedMedia file from the storage device.
+     *
+     * @param context A context that is used for opening a database connection.
+     * @param mediaId ID of the FeedMedia object whose downloaded file should be deleted.
+     */
+    public static Future<?> deleteFeedMediaOfItem(final Context context,
+                                                  final long mediaId) {
+        Future res = doDeleteFeedMediaOfItem(context, mediaId);
+        if (UserPreferences.shouldDeleteRemoveFromQueue()) {
+            DBWriter.removeQueueItem(context, DBReader.getFeedMedia(mediaId).getItem(), false);
```

Comment:
Short of making both DB operations done in the same transaction, one way to address the two asynchronous DB calls is to make the second call wait until the first (the actual deleteFeedMediaOfItem) is completed by waiting for the `Future`
```java
  res.get();
  DBWriter.removeQueueItem(...)
```

Reply:
I don't think that's a good idea. Then there is not really a point in using a future. Maybe one could create a new Future that calls both get() calls internally...

Reply:
@ByteHamster I am not sure what your concerns are exactly.

Functionally, making the second call wait for the first call ensures correctness: If for some reason the first call `deleteFeedMediaOfItem()` fails, the second call `removeQueueItem()` should not be called at all.

With the current code in the PR, the second call `removeQueueItem()` will proceed, potentially leading to weird result, e.g., an item is removed from the queue but the media still exists.

It is a valid use of `Future`: ensuring correct sequencing of asynchronous calls, as the two calls are not truly independent of one another.

Reply:
The point of a future is to have asynchronous tasks without blocking calls to `deleteFeedMediaOfItem`. If you do half of the work synchronously (by calling `get()` directly after creating the future), the method gets blocking.

Maybe we could do something like this, so it stays asynchronous:

        return dbExec.submit(() -> {
            res.get();
            DBWriter.removeQueueItem(...).get()
        });

I am not sure if this can lead to deadlocks though

Reply:
I see what you mean by not blocking the callers. 

I think the deadlock concern can be addressed by not using `dbExec`, given `dbExec` is a `SingleThreadExecutor`.
```java
//          V use some other executors other than dbExec, so that it won't block the dbExec.submit() call in DBWriter.removeQueueItem()
    return dbExec.submit(() -> { 
        res.get();
        DBWriter.removeQueueItem(...).get()
    });
```

Reply:
We currently only use one executor. I think creating a new executor just for this method could be a bit too much.

> With the current code in the PR, the second call removeQueueItem() will proceed, potentially leading to weird result, e.g., an item is removed from the queue but the media still exists.

To fix that, we need to move the `removeQueueItem()` call into the executor within doDeleteFeedMediaOfItem. The `get` function succeeds even if the media can not be deleted. Maybe we can create a private, synchronous version of `removeQueueItem` and just put it in the end of doDeleteFeedMediaOfItem.

Reply:
Yes, I agree we have to either: 
1. Use a new executor and call the underlying async `doDeleteFeedMediaOfItem()` and `removeQueueItem()` . I confirmed reusing the single-threaded `dbExec` will cause deadlock.

2. Refactor the codes so that there are private synchronous version of `doDeleteFeedMediaOfItem()` and `removeQueueItem()`, say, `doDeleteFeedMediaOfItemSync()` and `doRemoveQueueItemSync()`. The public asynchronous `deleteFeedMediaOfItem()` then uses `dbExec` call the 2 underlying sync versions.

Reply:
A slight variation of avoid a new executor is to use FutureTask
```java
    return new FutureTask(() -> {
        res.get();
        DBWriter.removeQueueItem(...).get()
    }, null);
```
It avoids creating another dedicated thread pool for the executor. It probably spawns a new thread each time the call is invoked. 

Given it's a relatively infrequent user-facing operation, rather than some low-level codes that might be used frequently. I think it's acceptable.

Reply:
Using a FutureTask does not make sure that the deletion succeeds before removing from queue

Reply:
> Using a FutureTask does not make sure that the deletion succeeds before removing from queue

We might have miscommunicated. What I mean is that 
- the underlying operations `doDeleteFeedMediaOfItem()` and `removeQueueItem()` remain to use `dbExec`
- the `.get()` of `doDeleteFeedMediaOfItem()` ensures deletion succeeds before removing it from the queue.
- a `FutureTask` is created to wrap the above operation and return the user, preventing deadlock that would have occured if the wrapping is done by `dbExec.submit()`

It'd be something like the following: 
```java
    FutureTask future = new FutureTask(() -> {
        doDeleteFeedMediaOfItem(context, mediaId).get();
        removeQueueItem(...).get();
    }, null);
    future.run();
    return future;
```

Reply:
> the .get() of doDeleteFeedMediaOfItem() ensures deletion succeeds before removing it from the queue.

I don't think so. The future submitted in `doDeleteFeedMediaOfItem` returns without throwing a exception if deletion went wrong

Reply:
`doDeleteFeedMediaOfItem()` will throw exception **when `.get()` is invoked** as above, unless `doDeleteFeedMediaOfItem()` swallows exceptions explicitly.

You can see it in a standalone test:
  https://gist.github.com/orionlee/366702f7efe56ed007db151a0fe1dfd6

Reply:
I have to amend the above statement. `doDeleteFeedMediaOfItem()`does silently returns if the actual file deletion fails. So for the above scheme to work, it needs to throw exception instead.
  https://github.com/AntennaPod/AntennaPod/blob/dfcb16d77b5e9b7ea3f5d19e528c9f4950eb4520/core/src/main/java/de/danoeh/antennapod/core/storage/DBWriter.java#L87